sgkit/distance/api.py

```diff
@@ -77,27 +77,53 @@ def pairwise_distance(
 
     >>> x = np.array([[6, 4, 1,], [4, 5, 2], [9, 7, 3]])
     >>> pairwise_distance(x, metric='correlation')
-    array([[1.11022302e-16, 2.62956526e-01, 2.82353505e-03],
-           [2.62956526e-01, 0.00000000e+00, 2.14285714e-01],
-           [2.82353505e-03, 2.14285714e-01, 0.00000000e+00]])
+    array([[-4.44089210e-16,  2.62956526e-01,  2.82353505e-03],
+           [ 2.62956526e-01,  0.00000000e+00,  2.14285714e-01],
+           [ 2.82353505e-03,  2.14285714e-01,  0.00000000e+00]])
     """
-
     try:
-        metric_ufunc = getattr(metrics, metric)
+        metric_map_func = getattr(metrics, f"{metric}_map")
+        metric_reduce_func = getattr(metrics, f"{metric}_reduce")
+        n_map_param = metrics.N_MAP_PARAM[metric]
     except AttributeError:
         raise NotImplementedError(f"Given metric: {metric} is not implemented.")
 
     x = da.asarray(x)
-    x_distance = da.blockwise(
-        # Lambda wraps reshape for broadcast
-        lambda _x, _y: metric_ufunc(_x[:, None, :], _y),
-        "jk",
+    if x.ndim != 2:
+        raise ValueError(f"2-dimensional array expected, got '{x.ndim}'")
+
+    def _pairwise(f, g):
+        result = metric_map_func(f[:, None, :], g, np.empty(n_map_param, dtype=x.dtype))
+        return result[..., np.newaxis]
+
+    out = da.blockwise(
+        _pairwise,
+        'ijk',
         x,
-        "ji",
+        'ik',
         x,
-        "ki",
-        dtype="float64",
-        concatenate=True,
+        'jk',
+        dtype=x.dtype,
+        concatenate=False,
     )
-    x_distance = da.triu(x_distance, 1) + da.triu(x_distance).T
-    return x_distance.compute()
+
+    def _aggregate(x_chunk, axis, keepdims):
+        # This is to handle the case when an empty array is passed in the first iteration
+        # This check is not required in dask >= 2021.1.1
+        if x_chunk.size != 0:
+            x_chunk = x_chunk.reshape(x_chunk.shape[:-2] + (-1, n_map_param))
```

Comment:
So this happens because dask needs to compute metadata for the reduction? Is this necessary? Is it possible/Should we specify `meta` for reduction to avoid this extra complexity? What's the common practice to deal with this?

Reply:
> So this happens because dask needs to compute metadata for the reduction?

I believe so.

> Is it possible/Should we specify meta for reduction to avoid this extra complexity? What's the common practice to deal with this?

Good suggestion. I think it should be possible to use meta, let me try that.